### PR TITLE
Issue #127 idempotent id processor

### PR DESCRIPTION
**Pull request by MiniPlayer**
First version of the processor. (Modified some Validator too and implemented some unit test on them)

### CODE REVIEW

File: logisland-plugins/logisland-common-processors-plugin/src/main/java/com/hurence/logisland/processor/ModifyId.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+    public static final AllowableValue GENERATE_RANDOM_UUID = new AllowableValue("randomUuid", "generate a random uid",
+            "generate a randomUid using java library");
+
+    public static final AllowableValue GENERATE_HASH = new AllowableValue("hashFields", "generate a hash from fields",
+            "generate a hash from fields");
+
+
+    public static final PropertyDescriptor STRATEGY = new PropertyDescriptor.Builder()
+            .name("strategy.to.use")
```
Comment (oalam) [inline logisland-plugins/logisland-common-processors-plugin/src/main/java/com/hurence/logisland/processor/ModifyId.java]: rename to "id.generation.strategy

File: logisland-plugins/logisland-common-processors-plugin/src/main/java/com/hurence/logisland/processor/ModifyId.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            .required(true)
+            .allowableValues(GENERATE_RANDOM_UUID, GENERATE_HASH)
+            .defaultValue(GENERATE_RANDOM_UUID.getValue())
+            .build();
+
+    /**
+     * properties sued only in case of Hash strategy
+     */
+    public static final PropertyDescriptor FIELDS_TO_USE_FOR_HASH = new PropertyDescriptor.Builder()
+            .name("fields.to.use.for.hash")
```
Comment (oalam) [inline logisland-plugins/logisland-common-processors-plugin/src/main/java/com/hurence/logisland/processor/ModifyId.java]: rename to "fields.to.hash"

File: logisland-plugins/logisland-common-processors-plugin/src/main/java/com/hurence/logisland/processor/ModifyId.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            .description("the comma separated list of field names (e.g. \"policyid,date_raw\"")
+            .required(true)
+            .addValidator(StandardValidators.COMMA_SEPARATED_LIST_VALIDATOR)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor CHARSET_TO_USE_FOR_HASH = new PropertyDescriptor.Builder()
+            .name("charset.to.use.for.hash")
```
Comment (oalam) [inline logisland-plugins/logisland-common-processors-plugin/src/main/java/com/hurence/logisland/processor/ModifyId.java]: rename to "hash.charset"

File: logisland-plugins/logisland-common-processors-plugin/src/main/java/com/hurence/logisland/processor/ModifyId.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            .description("the charset to use to hash id string (e.g. \"UTF-8\"")
+            .required(true)
+            .addValidator(StandardValidators.CHARACTER_SET_VALIDATOR)
+            .defaultValue("UTF-8")
+            .build();
+
+    public static final PropertyDescriptor HASH_ALGORITHM = new PropertyDescriptor.Builder()
+            .name("algorithm.to.use.for.hash")
```
Comment (oalam) [inline logisland-plugins/logisland-common-processors-plugin/src/main/java/com/hurence/logisland/processor/ModifyId.java]: rename to "hash.algorithm"

File: logisland-plugins/logisland-common-processors-plugin/src/main/java/com/hurence/logisland/processor/ModifyId.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            .description("the algorithme to use to hash id string (e.g. \"SHA-256\"")
+            .required(true)
+            .addValidator(StandardValidators.HASH_ALGORITHM_VALIDATOR)
+            .defaultValue("SHA-256")
```
Comment (oalam) [inline logisland-plugins/logisland-common-processors-plugin/src/main/java/com/hurence/logisland/processor/ModifyId.java]: add a list of allowable values

File: logisland-plugins/logisland-common-processors-plugin/src/main/java/com/hurence/logisland/processor/ModifyId.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    private static final long serialVersionUID = -270933070438408174L;
+
+    private static final Logger logger = LoggerFactory.getLogger(ModifyId.class);
+
+    public static final AllowableValue GENERATE_RANDOM_UUID = new AllowableValue("randomUuid", "generate a random uid",
+            "generate a randomUid using java library");
+
+    public static final AllowableValue GENERATE_HASH = new AllowableValue("hashFields", "generate a hash from fields",
+            "generate a hash from fields");
+
```
Comment (oalam) [inline logisland-plugins/logisland-common-processors-plugin/src/main/java/com/hurence/logisland/processor/ModifyId.java]: add a third strategy (TYPE_TIME_HASH) based on the pattern

$record_type-$record_time-hash($fields.to.hash)

File: logisland-plugins/logisland-common-processors-plugin/src/main/java/com/hurence/logisland/processor/ModifyId.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    /**
+     * properties sued only in case of Hash strategy
+     */
+    public static final PropertyDescriptor FIELDS_TO_USE_FOR_HASH = new PropertyDescriptor.Builder()
+            .name("fields.to.use.for.hash")
+            .description("the comma separated list of field names (e.g. \"policyid,date_raw\"")
+            .required(true)
+            .addValidator(StandardValidators.COMMA_SEPARATED_LIST_VALIDATOR)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
```
Comment (oalam) [inline logisland-plugins/logisland-common-processors-plugin/src/main/java/com/hurence/logisland/processor/ModifyId.java]: add default value with

Value_Raw_Content

File: logisland-plugins/logisland-common-processors-plugin/src/test/java/com/hurence/logisland/processor/ModifyIdTest.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        MockRecord outputRecord = testRunner.getOutputRecords().get(0);
+
+        outputRecord.assertRecordSizeEquals(4);
+        outputRecord.assertFieldEquals("string1",  "value1");
+        outputRecord.assertFieldEquals("string2",  "value2");
+        outputRecord.assertFieldEquals("long1",  1);
+        outputRecord.assertFieldEquals("long2",  2);
+        outputRecord.assertFieldEquals("record_id",  "��ok\u007Fwc��\"ݦ\u000B*\u000B��A*�\u0018K�N�x�M7�\u001D\u000B");
+    }
+}
```
Comment (oalam) [inline logisland-plugins/logisland-common-processors-plugin/src/test/java/com/hurence/logisland/processor/ModifyIdTest.java]: add a unit test to monitor performance overhead

should be less than x ms / record (loop over 10000 records for example)

File: logisland-plugins/logisland-common-processors-plugin/src/test/java/com/hurence/logisland/processor/ModifyIdTest.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+ */
+public class ModifyIdTest {
+    private Record getRecord1() {
+        Record record1 = new StandardRecord();
+        record1.setField("string1", FieldType.STRING, "value1");
+        record1.setField("string2", FieldType.STRING, "value2");
+        record1.setField("long1", FieldType.LONG, 1);
+        record1.setField("long2", FieldType.LONG, 2);
+        return record1;
+    }
```
Comment (oalam) [inline logisland-plugins/logisland-common-processors-plugin/src/test/java/com/hurence/logisland/processor/ModifyIdTest.java]: add test for other id generation strategy

File: logisland-plugins/logisland-common-processors-plugin/src/test/java/com/hurence/logisland/processor/ModifyIdTest.java (hunk 8 above)
Comment (oalam) [inline logisland-plugins/logisland-common-processors-plugin/src/test/java/com/hurence/logisland/processor/ModifyIdTest.java]: all options must be tested (charset for ex)

**Comment by MiniPlayer** [comment]
mmm, i think we should change travis build to show errors trace when build fail. Because I just restarted the build and it went from red to green... There seems to be something strange with current build. Would be interesting to have more details when it crashes.

**Pull request merged**